django_password_validators/password_history/password_validation.py

```diff
@@ -45,15 +48,32 @@ def validate(self, password, user=None):
 
         for user_config in UserPasswordHistoryConfig.objects.filter(user=user):
             password_hash = user_config.make_password_hash(password)
+
             try:
-                PasswordHistory.objects.get(
-                    user_config=user_config,
-                    password=password_hash
-                )
-                raise ValidationError(
-                    _("You can not use a password that is already used in this application."),
-                    code='password_used'
-                )
+                if self.lookup_range == 0:
+                    raise PasswordHistory.DoesNotExist
+
+                current_user_passwords = PasswordHistory.objects.filter(
+                    user_config=user_config).order_by('date')
+
+                if self.lookup_range >= len(current_user_passwords):
+                    PasswordHistory.objects.get(
+                        user_config=user_config,
+                        password=password_hash
+                    )
+                else:
+                    # At this point, there are passwords we have to delete
+                    for entry in current_user_passwords[:len(
+                            current_user_passwords) - self.lookup_range]:
+                        entry.delete()
+
+                    if any([entry.user_config == user_config and
+                            entry.password == password
+                            for entry in current_user_passwords[len(current_user_passwords) - self.lookup_range:]]):
```

Comment:
@spokerman12, `current_user_passwords` does not contain the right query results any more as you have iterated over it and deleted some of the items. So if you try to check the current password, you will be checking more than the configured number of lookup items. To prevent this, you will have to query the database again or use `.refresh_from_db()` to do it. 

Instead of doing this, you can drop the `if` condition and its block above, update the `else` block to an `if` condition with the negation of the original condition in the `if` block and use the existing lines 49-56 that you have removed in your PR as common code that will always execute. What do you think?

Instead of

Reply:
I spent a bit on that one! `.delete()` gets rid of the database entry, not the what's inside the variables. Similar to how we can create any users by doing `User(username='foo')`, they don't actually exist until they're deleted. 

```
>>> from django.contrib.auth.models import User
>>> p1 = User(username='Bob')
>>> p1.save()
>>> p2 = User(username='John')
>>> p2.save()
>>> p3 = User(username='Steve')
>>> p3.save()
>>> all_people = [p1,p2,p3]
>>> print(all_people)
[<User: Bob>, <User: John>, <User: Steve>]
>>> 
for p in all_people:
    if p.username == 'Bob':
        p.delete()
>>> ... ... ... 
(1, {'admin.LogEntry': 0, 'auth.User_groups': 0, 'auth.User_user_permissions': 0, 'auth.User': 1})
>>> print(all_people)
[<User: Bob>, <User: John>, <User: Steve>]
>>> User.objects.filter(username="Bob")                           
<QuerySet []>
```

Bob still exists in `p1`

Reply:
@spokerman12, 

> we can create any users by doing User(username='foo'), they don't actually exist until they're deleted.

Did you mean `saved` instead of `deleted`?

In the `else` block, which is executed only when `current_user_passwords` has more passwords than the lookup range, there will be older passwords which have to be deleted and shouldn't be used in the comparison.

But deleting the individual entries in `current_user_passwords` by iterating over them and calling the `.delete()` method on each of the candidate entries will not remove the entries from the `current_user_passwords` iterable.

So when you do the comparison using the values in `current_user_passwords` after deleting the older candidates, you will still be doing a comparison against the deleted passwords and that is not what we want. Please let me know if you have any questions.

Reply:
> Did you mean saved instead of deleted?
Yes! My bad.

> comparison against the deleted passwords and that is not what we want
Oh, note the `self.lookup_range:` at the end here https://github.com/fizista/django-password-validators/blob/28ac6eb533e9ae853568d941f1c6e24ff076516d/django_password_validators/password_history/password_validation.py#L72